lib/puppet/provider/chocolateyfeature/windows.rb

```diff
@@ -0,0 +1,138 @@
+require 'puppet/type'
+require 'pathname'
+require 'rexml/document'
+
+Puppet::Type.type(:chocolateyfeature).provide(:windows) do
+  confine :operatingsystem => :windows
+  defaultfor :operatingsystem => :windows
+
+  require Pathname.new(__FILE__).dirname + '../../../' + 'puppet_x/chocolatey/chocolatey_common'
+  include PuppetX::Chocolatey::ChocolateyCommon
+
+  MINIMUM_SUPPORTED_CHOCO_VERSION = '0.9.9.0'
+
+  commands :chocolatey => PuppetX::Chocolatey::ChocolateyCommon.chocolatey_command
+
+  def initialize(value={})
+    super(value)
+    @property_flush = {}
+  end
+
+  def properties
+    if @property_hash.empty?
+      @property_hash = query || { :ensure => ( :disabled )}
+      @property_hash[:ensure] = :disabled if @property_hash.empty?
+    end
+    @property_hash.dup
+  end
+
+  def query
+    self.class.features.each do |feature|
+      return feature.properties if @resource[:name][/\A\S*/].downcase == feature.name.downcase
+    end
+
+    return {}
+  end
+
+  def self.get_features
+    PuppetX::Chocolatey::ChocolateyCommon.set_env_chocolateyinstall
+
+    choco_config = PuppetX::Chocolatey::ChocolateyCommon.choco_config_file
+    raise Puppet::ResourceError, "Config file not found for Chocolatey. Please make sure you have Chocolatey installed." if choco_config.nil?
+    raise Puppet::ResourceError, "An install was detected, but was unable to locate config file at #{choco_config}." unless PuppetX::Chocolatey::ChocolateyCommon.file_exists?(choco_config)
+
+    Puppet.debug("Gathering features from '#{choco_config}'.")
+    config = REXML::Document.new File.new(choco_config, 'r')
+
+    config.elements.to_a( '//feature' )
+  end
+
+  def self.get_feature(element)
+    feature = {}
+    return feature if element.nil?
+
+    feature[:name]        = element.attributes['name'].downcase if element.attributes['name']
+    feature[:description] = element.attributes['description'].downcase if element.attributes['description']
+
+    enabled = false
+    enabled = element.attributes['enabled'].downcase == 'true' if element.attributes['enabled']
+
+    feature[:ensure] = :disabled
+    feature[:ensure] = :enabled if enabled
+
+    Puppet.debug("Loaded feature '#{feature.inspect}'.")
+
+    feature
+  end
+
+  def self.features
+    @features ||=  get_features.collect do |item|
```

Comment:
I'm not 100% sure why we're using a class level variable for features here.  The only other method that uses `@features` is `self.refresh_features` and that's only called by flush.

Should we scrap using the class level features variable and just query the filesystem every time?